themes/qgrid/source/js/menu.js

```diff
@@ -15,41 +15,196 @@ function toggleVisibility(e) {
+	if (!menuItem.querySelector('.tag')) {
+		menuItem.insertBefore(document.createElement('a'), menuItem.querySelector('.title'));
+		menuItem.querySelector('a').classList.add('tag');
+	}
+	const title = menuItem.querySelector('.title');
+	title.classList.add('menu-item');
+	if (!menuItem.querySelector('.border')) {
+		const border = menuItem.insertBefore(document.createElement('span'), menuItem.querySelector('.title'));
+		border.classList.add('border');
+		border.textContent = '/ ';
+	}
+	const menuTag = menuItem.querySelector('.tag');
+	menuTag.classList.remove('hidden');
+	menuTag.textContent = formatTag(tag, search);
+	menuTag.setAttribute('href', menuItem.querySelector('.title').getAttribute('href'));
+}
+
+function removeTagText(menuItem, search) {
+	const title = menuItem.querySelector('.title');
+	if (search) {
+		menuItem.classList.add('hidden');
+	}
+	else {
+		menuItem.classList.remove('hidden');
+	}
+	const tag = menuItem.querySelector('.tag');
+	if (tag) {
+		tag.classList.add('hidden');
+	}
+	title.classList.remove('menu-item');
+	if (menuItem.querySelector('.border')) {
+		menuItem.removeChild(menuItem.querySelector('.border'));
+	}
+}
+
+function highlightText(item, search) {
+	const { textContent } = item;
+	const searchContains = new RegExp(search, 'ig');
+	if (search && searchContains.test(textContent)) {
+		item.innerHTML = textContent.replace(searchContains, elem => `<span class="highlight">${elem}</span>`);
+		item.parentElement.classList.remove('hidden');
+		return true;
+	}
+	item.innerHTML = textContent;
+}
+
+function formatTag(tag, search) {
+	let tagsContent = split(tag);
+	const index = getIndexOfTag(tagsContent, search);
+	const visibleWordCount = 2;
+	if (tagsContent.length > visibleWordCount) {
+		if (index < tagsContent.length - visibleWordCount) {
+			tagsContent.splice(index + visibleWordCount, tagsContent.length, ' ...');
+		}
+		if (index > 0) {
+			tagsContent.splice(0, index, '...');
+		}
+	}
+	return tagsContent.join('');
+}
+
+function getIndexOfTag(tag, search) {
+	const arraySearch = split(search);
+	for (let indexTag = 0; indexTag < tag.length; indexTag++) {
+		let indexSearch = 0;
+		let index = tag[indexTag].toLowerCase().indexOf(arraySearch[indexSearch]);
+		while (index >= 0) {
+			if (indexSearch == arraySearch.length - 1) {
+				return indexTag;
 			}
+			indexSearch++;
+			index = tag[indexSearch + indexTag].toLowerCase().indexOf(arraySearch[indexSearch]);
 		}
 	}
 }
 
+function updateMenuLinks(search) {
+	const nav = document.getElementById('nav');
+	const hash = search ? '#search=' + search : '';
+	for (let title of nav.querySelectorAll('.title')) {
+		let href = title.getAttribute('href');
+		const index = href.indexOf('#search=');
+		href = (index >= 0) ? href.slice(0, index) : href;
+		title.setAttribute('href', href + hash);
+	}
+}
+
+function split(text) {
+	const template = new RegExp(/\s?[a-z,A-Z,0-9,&,-]+/, 'g');
+	return text.match(template) || text.match(/\s/, 'g');
```

Comment:
why we need second match?

Reply:
lest the function should return null, when we enter a space in the search. It can be done through the replacing `+` with `*` in regexp, but then this function will be return an array with an extra empty element